### PR TITLE
get configs from service

### PR DESCRIPTION
**Pull request by yaronpx**
* Refactored and created a context object
* Get configurations dynamically from configuration service

### CODE REVIEW

File: lib/perimeterx.js [hunk 1]
```diff
@@ -2,8 +2,9 @@
 
 const pxClient = require('./utils/pxclient');
 const pxlogger = require('./utils/pxlogger');
+const PxContext = require('./utils/pxcontext');
+const PxConfig = require('./pxconfig');
 const mu = require('mu2');
-var htmlEncode = require('htmlencode').htmlEncode;
```
Comment (pxjohnny) [inline lib/perimeterx.js]: htmlEncode is no longer used in the project, please remove it from package.json as well...

File: lib/utils/configloader.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                this.config.checksum = body.checksum;
+                this.config.COOKIE_SECRET_KEY = body.cookieKey;
+                this.config.PX_APP_ID = body.appId;
+                this.config.BLOCKING_SCORE = body.blockingScore;
+                this.config.DEBUG_MODE = body.debugMode;
+                this.config.ENABLE_MODULE = body.moduleEnabled;
+                this.config.SENSITIVE_HEADERS = body.sensitiveHeaders;
+                this.config.ACTIVITIES_TIMEOUT = body.connectTimeout;
+                this.config.API_TIMEOUT_MS = body.riskTimeout;
+                this.config.blockPageTemplate = body.blockPageTemplate;
```
Comment (nitzanpx) [inline lib/utils/configloader.js]: remove templates meanwhile

File: lib/utils/pxclient.js [hunk 3]
```diff
@@ -64,7 +65,7 @@ function submitActivities() {
     const headers = {
         'Content-Type': 'application/json'
     };
-    pxhttpc.callServer(tempActivities, headers, pxconfig.SERVER_COLLECT_URI, 'activities', (err) => {
+    pxhttpc.callServer(tempActivities, headers, pxConfig.conf.SERVER_COLLECT_URI, 'activities', (err) => {
```
Comment (nitzanpx) [inline lib/utils/pxclient.js]: not sure, but should this `pxConfig` be `config`?`

File: lib/utils/pxhttpc.js [hunk 4]
```diff
@@ -15,16 +15,17 @@ exports.callServer = callServer;
...
         body: JSON.stringify(data),
         headers: headers,
-        proxy: pxconfig.PROXY_URL,
+        proxy: config.PROXY_URL,
         agent: keepaliveAgent
     };
 
     if (callType === 'query') {
-        callData.timeout = pxconfig.API_TIMEOUT_MS
+        callData.timeout = callType === 'query' ? config.API_TIMEOUT_MS : config.ACTIVITIES_TIMEOUT;
```
Comment (nitzanpx) [inline lib/utils/pxhttpc.js]: remove the if statement or set the default timeout above

File: lib/utils/pxclient.js (hunk 3 above)
Comment (yaronpx) [inline lib/utils/pxclient.js]: no, that's ok

File: lib/utils/pxhttpc.js [hunk 5]
```diff
@@ -71,7 +72,7 @@ function getHttpAgent() {
     return new HttpAgent({
         maxSockets: 100,
         maxFreeSockets: 10,
-        timeout: pxconfig.API_TIMEOUT_MS,
+        timeout: pxConfig.conf.API_TIMEOUT_MS,
```
Comment (nitzanpx) [inline lib/utils/pxhttpc.js]: is possible to use `config` instead of `pxConfig`?

File: lib/utils/pxlogger.js [hunk 6]
```diff
@@ -3,15 +3,15 @@
  * Version 1.0 Published 12 May 2016
  */
 
-const pxconfig = require('../pxconfig').conf();
+const pxconfig = require('../pxconfig');
```
Comment (nitzanpx) [inline lib/utils/pxlogger.js]: can we adjust the naming convention to pxConfig?

File: lib/utils/pxutil.js [hunk 7]
```diff
@@ -29,7 +27,7 @@ exports.verifyDefined = verifyDefined;
  * @param {Function} callback - callback function.
  */
 function verifyUserScore(pxCtx, callback) {
-    const pxconfig = require('../pxconfig').conf();
+    const pxconfig = require('../pxconfig').conf;
```
Comment (nitzanpx) [inline lib/utils/pxutil.js]: naming convention

File: lib/utils/pxutil.js [hunk 8]
```diff
@@ -76,15 +74,15 @@ function verifyUserScore(pxCtx, callback) {
  * @return {Array} request headers an array format.
  */
 function formatHeaders(headers) {
-    const pxconfig = require('../pxconfig').conf();
+    const pxconfig = require('../pxconfig').conf;
```
Comment (nitzanpx) [inline lib/utils/pxutil.js]: naming convention

File: lib/utils/pxutil.js [hunk 9]
```diff
@@ -122,7 +120,7 @@ function checkForStatic(req, exts) {
  */
 function filterSensitiveHeaders(headers) {
     try {
-        const pxconfig = require('../pxconfig').conf();
+        const pxconfig = require('../pxconfig').conf;
```
Comment (nitzanpx) [inline lib/utils/pxutil.js]: naming convention

File: lib/utils/pxapi.js [hunk 10]
```diff
@@ -116,16 +118,16 @@ function evalByServerCall(pxCtx, callback) {
...
-    if (!res || !pxutil.verifyDefined(res.score) || !res.action) {
+    if (!res || !pxUtil.verifyDefined(res.score) || !res.action) {
         return -1;
     }
     let score = res.score;
-    pxCtx.block_score = score;
-    pxCtx.block_action = res.action;
-    pxCtx.uuid = res.uuid;
-    if (score >= pxconfig.BLOCKING_SCORE) {
+    if (score >= pxConfig.conf.BLOCKING_SCORE) {
```
Comment (nitzanpx) [inline lib/utils/pxapi.js]: if possible to align , use`config` instead

Comment (yaronpx) [inline lib/utils/pxapi.js]: in this case it's not really necessary, since it's the only use of it

File: lib/perimeterx.js [hunk 11]
```diff
@@ -2,8 +2,9 @@
 
 const pxClient = require('./utils/pxclient');
 const pxlogger = require('./utils/pxlogger');
```
Comment (nitzanpx) [inline lib/perimeterx.js]: possible to align the name pxLogger if already started on different variables

File: lib/perimeterx.js [hunk 12]
```diff
@@ -129,7 +130,7 @@ function pxBlock(pxCtx, action, customHandler, req, res, next) {
             block_reason: reason,
             client_uuid: pxCtx.uuid,
             block_module: 'px-node-express',
```
Comment (nitzanpx) [inline lib/perimeterx.js]: not related to this pr but what is this?

Comment (yaronpx) [inline lib/perimeterx.js]: I guess it's metadata so we can have this data in ES

File: lib/pxconfig.js [hunk 13]
```diff
@@ -70,56 +66,71 @@ PX_DEFAULT.CUSTOM_LOGO = '';
 PX_DEFAULT.TEMPLATE = 'captcha';
 PX_DEFAULT.LOGO_VISIBILITY = 'hidden';
 PX_DEFAULT.SENSITIVE_ROUTES = [];
+PX_DEFAULT.DYNAMIC_CONFIGURATIONS = false;
+PX_DEFAULT.CONFIGURATION_LOAD_INTERVAL = 5000;
+PX_DEFAULT.configurationServer = 'https://px-conf-mgmt-stg.perimeterx.net/module';
```
Comment (nitzanpx) [inline lib/pxconfig.js]: break path to put it in PX_INTERNAL like examples above

**Pull request merged**